File: swift/BCCrypto/Sources/BCCrypto/PublicKeyEncryption.swift
import BCRand
import CryptoKit
import Foundation

private let genericPrivateKeySize = 32
private let genericPublicKeySize = 32

/// The size in bytes of an X25519 private key.
public let x25519PrivateKeySize = 32

/// The size in bytes of an X25519 public key.
public let x25519PublicKeySize = 32

/// Derives an X25519 agreement private key from arbitrary key material using HKDF.
///
/// - Parameter keyMaterial: The input key material.
/// - Returns: A 32-byte derived private key suitable for key agreement.
public func deriveAgreementPrivateKey(_ keyMaterial: Data) -> Data {
    hkdfHmacSHA256(
        keyMaterial: keyMaterial,
        salt: Data("agreement".utf8),
        keyLength: genericPrivateKeySize
    )
}

/// Derives a signing private key from arbitrary key material using HKDF.
///
/// - Parameter keyMaterial: The input key material.
/// - Returns: A 32-byte derived private key suitable for signing.
public func deriveSigningPrivateKey(_ keyMaterial: Data) -> Data {
    hkdfHmacSHA256(
        keyMaterial: keyMaterial,
        salt: Data("signing".utf8),
        keyLength: genericPublicKeySize
    )
}

/// Generates a new random X25519 private key.
///
/// - Parameter rng: The random number generator to use.
/// - Returns: A 32-byte private key.
public func x25519NewPrivateKeyUsing<R: BCRandomNumberGenerator>(
    _ rng: inout R
) -> Data {
    rng.randomData(count: x25519PrivateKeySize)
}

/// Derives the X25519 public key from a private key.
///
/// - Parameter privateKey: A 32-byte X25519 private key.
/// - Returns: The corresponding 32-byte public key.
public func x25519PublicKeyFromPrivateKey(_ privateKey: Data) -> Data {
    requireLength(
        privateKey,
        expected: x25519PrivateKeySize,
        name: "privateKey"
    )
    let key = try! Curve25519.KeyAgreement.PrivateKey(
        rawRepresentation: privateKey
    )
    return key.publicKey.rawRepresentation
}

/// Computes a shared symmetric key from an X25519 key pair using Diffie-Hellman
/// key agreement followed by HKDF derivation.
///
/// - Parameters:
///   - privateKey: The local 32-byte X25519 private key.
///   - publicKey: The remote 32-byte X25519 public key.
/// - Returns: A 32-byte derived symmetric key.
public func x25519SharedKey(
    privateKey: Data,
    publicKey: Data
) -> Data {
    requireLength(
        privateKey,
        expected: x25519PrivateKeySize,
        name: "privateKey"
    )
    requireLength(
        publicKey,
        expected: x25519PublicKeySize,
        name: "publicKey"
    )

    let x25519Private = try! Curve25519.KeyAgreement.PrivateKey(
        rawRepresentation: privateKey
    )
    let x25519Public = try! Curve25519.KeyAgreement.PublicKey(
        rawRepresentation: publicKey
    )
    let secret = try! x25519Private.sharedSecretFromKeyAgreement(with: x25519Public)
    let shared = secret.withUnsafeBytes { Data($0) }

    return hkdfHmacSHA256(
        keyMaterial: shared,
        salt: Data("agreement".utf8),
        keyLength: symmetricKeySize
    )
}

File: go/bccrypto/public_key_encryption.go
package bccrypto

import (
	bcrand "github.com/nickel-blockchaincommons/bcrand-go"
	"golang.org/x/crypto/curve25519"
)

const (
	// GenericPrivateKeySize is the byte length of a generic private key.
	GenericPrivateKeySize = 32
	// GenericPublicKeySize is the byte length of a generic public key.
	GenericPublicKeySize = 32
	// X25519PrivateKeySize is the byte length of an X25519 private key.
	X25519PrivateKeySize = 32
	// X25519PublicKeySize is the byte length of an X25519 public key.
	X25519PublicKeySize = 32
)

// DeriveAgreementPrivateKey derives a 32-byte agreement private key from key material.
func DeriveAgreementPrivateKey(keyMaterial []byte) [GenericPrivateKeySize]byte {
	derived := HKDFHMACSHA256(keyMaterial, []byte("agreement"), GenericPrivateKeySize)
	var out [GenericPrivateKeySize]byte
	copy(out[:], derived)
	return out
}

// DeriveSigningPrivateKey derives a 32-byte signing private key from key material.
func DeriveSigningPrivateKey(keyMaterial []byte) [GenericPrivateKeySize]byte {
	derived := HKDFHMACSHA256(keyMaterial, []byte("signing"), GenericPrivateKeySize)
	var out [GenericPrivateKeySize]byte
	copy(out[:], derived)
	return out
}

// X25519NewPrivateKeyUsing creates a new X25519 private key using rng.
func X25519NewPrivateKeyUsing(rng bcrand.RandomNumberGenerator) [X25519PrivateKeySize]byte {
	data := rng.RandomData(X25519PrivateKeySize)
	var out [X25519PrivateKeySize]byte
	copy(out[:], data)
	return out
}

// X25519PublicKeyFromPrivateKey derives an X25519 public key from a private key.
func X25519PublicKeyFromPrivateKey(
	privateKey [X25519PrivateKeySize]byte,
) [X25519PublicKeySize]byte {
	publicKey, err := curve25519.X25519(privateKey[:], curve25519.Basepoint)
	if err != nil {
		panic("bccrypto: invalid X25519 private key")
	}
	var out [X25519PublicKeySize]byte
	copy(out[:], publicKey)
	return out
}

// X25519SharedKey computes the shared symmetric key from private/public key pair.
func X25519SharedKey(
	privateKey [X25519PrivateKeySize]byte,
	publicKey [X25519PublicKeySize]byte,
) [SymmetricKeySize]byte {
	sharedSecret, err := curve25519.X25519(privateKey[:], publicKey[:])
	if err != nil {
		panic("bccrypto: invalid X25519 key material")
	}
	derived := HKDFHMACSHA256(sharedSecret, []byte("agreement"), SymmetricKeySize)
	var out [SymmetricKeySize]byte
	copy(out[:], derived)
	return out
}
